### PR TITLE
Refactor: Shadow Models

### LINKED ISSUE #5: Folder structure
# Feature
## Desired Behavior / Functionality
The folder structure should be clean when it comes to storing models and data.

## What Needs to Be Done
Come up with a systematic way to store datasets, store models, and store shadow models.

## How Can It Be Tested
Inspection.

### LINKED ISSUE #24: Refactoring: Shadow models
# Feature
## Desired Behavior / Functionality
The shadow models should be reused among mutliple attacks and be within their own class object. 
Further, we should leverage composition (https://stackify.com/oop-concepts-composition/)

## What Needs to Be Done
- Bring shadow models to their own class.
- Follow the composition pattern for the attacks leveraging the shadow model.
- Store and read shadow models to ensure reusability.

## How Can It Be Tested
- Unit tests
- Run the attacks and make sure they still work

### PR DESCRIPTION
**Pull request by johanos1**
# Description

Refactoring of the code base

Summary of changes

* moved the code related to target model into dev_utils and config files into dev_config folder.
* Updated stored metadata to include initialization parameters to be able to create shadow models directly from the target model and its metadata.
* Removed attack_objects.py and attack_utils.py
* Created a singleton class: ShadowHandler that is responsible for managing shadow models. All attacks may access this class to access shadow models stored in the attack_objects folder.
* Updated config to provide parameters specific for each attack.
* Updated shadow model creation and config file to allow for dynamic creation of shadow models, optimizers, and loss functions.
* Updated attacks to follow a composition pattern.
* Extensive usage of logger
* Extensive checks of variable values
* 

## Resolved Issues

- [ ] fixes #24 
- [ ] fixes #5 

## How Has This Been Tested?
RMIA, P, and QMIA attacks are running.

### CODE REVIEW

File: leakpro/attacks/utils/shadow_model_handler.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+                labels = labels.long()  # noqa: PLW2901
+                inputs, labels = inputs.to(gpu_or_cpu, non_blocking=True), labels.to(gpu_or_cpu, non_blocking=True)  # noqa: PLW2901
+                optimizer.zero_grad()
+                outputs = shadow_model(inputs)
+                loss = criterion(outputs, labels)
+                pred = outputs.data.max(1, keepdim=True)[1]
+                loss.backward()
+
+                # Check if gradients are not zero
+                gradients = {name: param.grad for name, param in shadow_model.named_parameters()}
```
Comment (fazelehh) [inline leakpro/attacks/utils/shadow_model_handler.py:229]: Should we keep this check?

File: leakpro/attacks/utils/attack_data.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    """
+    if population_size <= 0:
+        raise ValueError("Population size must be greater than 0.")
+    if train_indices is None:
+        raise ValueError("Train indices must be provided.")
+    if test_indices is None:
+        raise ValueError("Test indices must be provided.")
+
+    all_index = np.arange(population_size)
+    not_allowed_indices = train_indices if sample_with_test_data else np.concatenate((train_indices, test_indices), axis=0)
```
Comment (fazelehh) [inline leakpro/attacks/utils/attack_data.py:37]: The variable name( attack_type_to_factory) is vague and probably should be the other way around here.

File: leakpro/reporting/utils.py [hunk 3]
```diff
@@ -28,30 +25,21 @@ def prepare_priavcy_risk_report(
...
+        raise ValueError("Please provide the audit results")
+
+    # Generate privacy risk report for auditing the model
+    ROCCurveReport.generate_report(
+        metric_result=audit_results,
+        save=True,
+        filename=f"{save_path}/ROC.png",
+        configs=configs,
+    )
+    SignalHistogramReport.generate_report(
```
Comment (fazelehh) [inline leakpro/reporting/utils.py:41]: Adding a report generator function to abstract_mia and calling whatever type of report we want from each attack itslef.

**Review by fazelehh — APPROVED** [review]
Great work! Added some minor comments.

File: leakpro/attacks/attack_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            target_model (torch.nn.Module): The target model.
+            target_model_metadata (Dict[str, Any]): The metadata of the target model.
+            configs (Dict[str, Any]): The configurations.
+            logger (logging.Logger): The logger object.
+
+        """
+        if configs["audit"]["attack_type"] not in list(self.attack_type_to_factory.keys()):
+            raise ValueError(
+                f"Unknown attack type: {configs['audit']['attack_type']}. "
+                f"Supported attack types: {self.attack_types}"
```
[the remark this answers is not in the document]
Comment (johanos1) [reply leakpro/attacks/attack_scheduler.py:39]: Done.

File: leakpro/attacks/utils/shadow_model_handler.py (hunk 1 above)
Comment (johanos1) [reply leakpro/attacks/utils/shadow_model_handler.py:229]: No, it should go. Good eye!

File: leakpro/attacks/utils/attack_data.py (hunk 2 above)
Comment (johanos1) [reply leakpro/attacks/utils/attack_data.py:37]: You mean sample_with_test_data? I replaced it with `test_included_in_auxiliary_data`

File: leakpro/reporting/utils.py (hunk 3 above)
Comment (johanos1) [reply leakpro/reporting/utils.py:41]: I think this should be its own issue. I created it as issue#35
